crates/subspace-fraud-proof/src/valid_bundle_proof.rs

```diff
@@ -0,0 +1,201 @@
+{
+    fn clone(&self) -> Self {
+        Self {
+            consensus_client: self.consensus_client.clone(),
+            executor: self.executor.clone(),
+            _phantom: self._phantom,
+        }
+    }
+}
+
+impl<CBlock, DomainBlock, CClient, Exec>
+    ValidBundleProofVerifier<CBlock, DomainBlock, CClient, Exec>
+where
+    CBlock: BlockT,
+    DomainBlock: BlockT,
+    CClient: HeaderBackend<CBlock> + ProvideRuntimeApi<CBlock> + Send + Sync,
+    CClient::Api: DomainsApi<CBlock, NumberFor<DomainBlock>, <DomainBlock as BlockT>::Hash>,
+    Exec: CodeExecutor + RuntimeVersionOf + 'static,
+{
+    /// Constructs a new instance of [`ValidBundleProofVerifier`].
+    pub fn new(consensus_client: Arc<CClient>, executor: Arc<Exec>) -> Self {
+        Self {
+            consensus_client,
+            executor,
+            _phantom: Default::default(),
+        }
+    }
+
+    /// Derive the `bundle_digest` of the given `bundle`
+    fn compile_bundle_digest(
+        &self,
+        domain_runtime_code: Vec<u8>,
+        bundle: OpaqueBundleOf<DomainBlock, CBlock, Balance>,
+    ) -> Result<H256, VerificationError> {
+        let mut extrinsics = Vec::with_capacity(bundle.extrinsics.len());
+        for opaque_extrinsic in bundle.extrinsics {
+            let ext = <<DomainBlock as BlockT>::Extrinsic>::decode(
+                &mut opaque_extrinsic.encode().as_slice(),
+            )?;
+            extrinsics.push(ext);
+        }
+
+        let runtime_api_light =
+            RuntimeApiLight::new(self.executor.clone(), domain_runtime_code.into());
+        let bundle_digest: Vec<_> =
+            <RuntimeApiLight<Exec> as DomainCoreApi<DomainBlock>>::extract_signer(
+                &runtime_api_light,
+                // `extract_signer` is a stateless runtime api thus it is okay to use
+                // default block hash
+                Default::default(),
+                extrinsics,
+            )?
+            .into_iter()
+            .map(|(signer, tx)| (signer, BlakeTwo256::hash_of(&tx)))
+            .collect();
+
+        Ok(BlakeTwo256::hash_of(&bundle_digest))
+    }
+
+    /// Verify the `ValidBundleProof`
+    pub fn verify(
+        &self,
+        valid_bundle_proof: &ValidBundleProof<
+            NumberFor<CBlock>,
+            CBlock::Hash,
+            NumberFor<DomainBlock>,
+            <DomainBlock as BlockT>::Hash,
+            Balance,
+        >,
+    ) -> Result<(), VerificationError> {
+        let consensus_block_header = {
+            let consensus_block_hash = valid_bundle_proof.bad_receipt.consensus_block_hash;
+            self.consensus_client
+                .header(consensus_block_hash)?
+                .ok_or_else(|| {
+                    sp_blockchain::Error::Backend(format!(
+                        "Header for {consensus_block_hash} not found"
+                    ))
+                })?
+        };
+        let parent_consensus_block_header = {
```

Comment:
As discussed in the call, this is not the way to go. If the runtime_api is returning something wrong then it should be fixed rather we making patches around it. This essentially makes a single fraud proof dependent on multiple consensus block states rather than the state of the block from which ER is derived. This also makes other usages of this API dependent on multiple consensus blocks. Either fix the issue in this PR or in seperate PR and adapt the code here accordnlgy

Reply:
> If the runtime_api is returning something wrong then it should be fixed rather we making patches around it.

Well, it is fixed here properly with the `parent_consensus_hash`, maybe just not the way you wanted, anyway will open a dedicated issue for this discussion and post it here.

> This essentially makes a single fraud proof dependent on multiple consensus block states rather than the state of the block from which ER is derived. This also makes other usages of this API dependent on multiple consensus blocks.

Why do we introduce such an assumption/requirement/limit to fraud proof, is it really necessary?